test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java

```diff
@@ -222,6 +245,68 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext contex
         LOG.debugf("postProcessTestInstance: %s ms", TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start));
     }
 
+    @Override
+    public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+            throws ParameterResolutionException {
+        // 1. Only test methods are supported
+        // 2. A param annotated with @SkipInject is never supported
+        if (parameterContext.getTarget().isPresent()
+                && !parameterContext.isAnnotated(SkipInject.class)
+                && parameterContext.getDeclaringExecutable().isAnnotationPresent(Test.class)) {
+            java.lang.reflect.Type requiredType = parameterContext.getParameter().getParameterizedType();
+            if (!requiredType.equals(TestInfo.class)) {
+                BeanManager beanManager = Arc.container().beanManager();
+                Annotation[] qualifiers = getQualifiers(parameterContext.getAnnotatedElement(), beanManager);
+                if (qualifiers.length > 0 && Arrays.stream(qualifiers).anyMatch(All.Literal.INSTANCE::equals)) {
+                    // @All List<>
+                    if (isListRequiredType(requiredType)) {
+                        return true;
+                    } else {
+                        throw new IllegalStateException("Invalid injection point type: " + parameterContext.getParameter());
+                    }
+                } else {
+                    // Only resolve the parameter if a single matching bean exists
```

Comment:
IIUIC, in this case you'll leave unresolvable parameters unclaimed which will lead to a junit exception when the test is executed.
However, that exception doesn't tell user much in terms of ambiguous/unsatisfied dependency and I think a lot of setups initially have some beans missing, accidentally removed, ambiguous etc. and seeing the underlying CDI exception might be more helpful. That's also one of the reasons why Weld-junit is greedy and claims all parameters by default.

Reply:
I think it would actually be best to `return false` if the dependency is unsatisfied, because that is the only general solution for allowing other `ParameterResolver`s to exist. It would also allow removing the extra check for `TestInfo`.

If the dependency is ambiguous ... I have no idea :-)

Reply:
As an extra point, if the parameter declares any qualifier, an unsatisfied dependency should be treated similarly to how an ambiguous dependency is treated (which I'm not sure what's the best), because that is clearly an indicator that the user expected a CDI lookup.

Reply:
>  It would also allow removing the extra check for TestInfo.

I'd like to keep the extra check and extend it according to https://github.com/quarkusio/quarkus/pull/38379#discussion_r1466241879. It does not make sense to lookup those types at all.

> If the dependency is ambiguous ... I have no idea :-)

We could log an ERROR if an ambiguous dependency/an unsatisfied dependency with a qualifier exists. But I don't think it's a good idea to "force inject" all params, i.e. `supportsParameter()` always returns `true`.

Reply:
> I think it would actually be best to `return false` if the dependency is unsatisfied, because that is the only general solution for allowing other `ParameterResolver`s to exist. It would also allow removing the extra check for `TestInfo`.

Even in case it's ambiguous for CDI, you could have another extension providing that parameter. For instance based on some non-CDI annotation present on the one of the two different impl classes.

> As an extra point, if the parameter declares any qualifier, an unsatisfied dependency should be treated similarly to how an ambiguous dependency is treated (which I'm not sure what's the best), because that is clearly an indicator that the user expected a CDI lookup.

You could do this for both, unsatisfied and ambiguous dependencies. So long as there is CDI qualifier and CDI cannot resolve it, you know it's an issue.
However, I am not sure this is just as clear to a user as it is to us. You will introduce a semi-magical behavior where *some* parameters are reported as problems with your CDI setup whereas others will just end up being reported as Junit not resolving them while the underlying issue is the same. So I guess I'd be -1 on that.

Plus a lot a of tests operate on simple beans without explicit qualifiers (i.e. they just have built-in `@Default`, `@Any`) so they will immediately fall into the latter category...

Reply:
You can only hard-code a limited set of parameters that you know upfront, and you don't have to attempt to resolve those to be able to `return false`. Once you're past that, you still have to decide whether an unsatisfied dependency (where the parameter didn't declare any qualifier) is truly a missing bean, or is supposed to be processed by a `ParameterResolver` you don't know about...

Reply:
It's a tough call. If only JUnit had a notion of parameter resolver priority :shrug:

Reply:
Well, JUnit currently requires exactly one extension resolving each parameter. It's an error if there is zero or more than one.
So we're looking for a middle ground between giving user as much information as we can on what's wrong with their setup (CDI-wise) and not clashing with other potential extensions trying to resolve same params.

Another reason why I am leaning towards greedy approach is that once you clash with another extension, JUnit will actually tell you what is the contested parameter and which exact extensions are trying to claim it - this makes it pretty easy to spot what's wrong and take appropriate steps (in this case use the Quarkus annotation on given parameter).
The other scenario is IMO harder to resolve as you are just told there is no extension resolving the parameter with no hint about given param being ambig/unsatisfied bean.

Reply:
Well, I don't want to make it complicated.

I think that we should:
- obviously inject only `@Test` methods
- automatically skip some hard-coded builtin params (`TestInfo` etc.)
- never inject a param/method annotated with `@SkiptInject`
- unsatisfied dependency 
  - qualifier or `@InjectMock` is declared -> throw an exception, because it's very likely a user/config error
  - otherwise log INFO (and mention `@SkipInject`) and supported=false
- ambiguous dependency -> throw an exception, if there's some other extension that resolves the param then it's clear that `@SkipInject` is needed

This is IMO reasonable compromise that should work in most cases and provides some useful info for users if something goes wrong.

Reply:
That looks sensible to me.

Reply:
Fair enough, I am still a fan of the greedy approach but this is a good compromise :⁠-⁠)